### ISSUE TITLE
"freeimgs" part of do_free_game always causes segmentation fault

### ISSUE THREAD

**Issue by ZeanZarzin** [aautin/42_so_long#4]
This error goes a bit deeper than the others:
`do_free_game` tries to use `free` on `game->imgs[i].name`:

line 32 in `utils.c`:
```c
free(game->imgs[i].name);
```

However, `game->imgs[i].name` is set using `ft_strnstr`, so it's somewhere within an area `malloc` has allocated:

line 35 in `src/image.c`:
```c
game->imgs[i].name = ft_strnstr(imgs_name[i], "sprites", len);
```
`imgs_name[i]` is the `char*` given by `malloc`, so `game->imgs[i].name` just points to somewhere within that allocated area,
but `free` only accepts the pointers directly given by `malloc`.

In other words, `free` only works when you give it the location of the *beginning* of an allocated area.
Because of this, using `free` on `game->imgs[i].name` will cause a segmentation fault.

Currently this isn't really a big deal because `do_free_game` with `freeimgs` is only used when the program is meant to terminate, but if you were to use it in the middle of the program somewhere or before saving, it would cause problems.

**Comment by aautin** [comment]
YEAH! I commited and pushed that yesterday at the end of the day, keeping in head that I have to fix that today!
I think I'll strdup each result of ft_strnstr so I can free the entire char **imgs_name and don't have leaks issues anymore.

**Comment by aautin** [comment]
> This error goes a bit deeper than the others: `do_free_game` tries to use `free` on `game->imgs[i].name`:
> 
> line 32 in `utils.c`:
> 
> ```c
> free(game->imgs[i].name);
> ```
> 
> However, `game->imgs[i].name` is set using `ft_strnstr`, so it's somewhere within an area `malloc` has allocated:
> 
> line 35 in `src/image.c`:
> 
> ```c
> game->imgs[i].name = ft_strnstr(imgs_name[i], "sprites", len);
> ```
> 
> `imgs_name[i]` is the `char*` given by `malloc`, so `game->imgs[i].name` just points to somewhere within that allocated area, but `free` only accepts the pointers directly given by `malloc`.
> 
> In other words, `free` only works when you give it the location of the _beginning_ of an allocated area. Because of this, using `free` on `game->imgs[i].name` will cause a segmentation fault.
> 
> Currently this isn't really a big deal because `do_free_game` with `freeimgs` is only used when the program is meant to terminate, but if you were to use it in the middle of the program somewhere or before saving, it would cause problems.

**Comment by aautin** [comment]
> This error goes a bit deeper than the others: `do_free_game` tries to use `free` on `game->imgs[i].name`:
> 
> line 32 in `utils.c`:
> 
> ```c
> free(game->imgs[i].name);
> ```
> 
> However, `game->imgs[i].name` is set using `ft_strnstr`, so it's somewhere within an area `malloc` has allocated:
> 
> line 35 in `src/image.c`:
> 
> ```c
> game->imgs[i].name = ft_strnstr(imgs_name[i], "sprites", len);
> ```
> 
> `imgs_name[i]` is the `char*` given by `malloc`, so `game->imgs[i].name` just points to somewhere within that allocated area, but `free` only accepts the pointers directly given by `malloc`.
> 
> In other words, `free` only works when you give it the location of the _beginning_ of an allocated area. Because of this, using `free` on `game->imgs[i].name` will cause a segmentation fault.
> 
> Currently this isn't really a big deal because `do_free_game` with `freeimgs` is only used when the program is meant to terminate, but if you were to use it in the middle of the program somewhere or before saving, it would cause problems.

YEAH! I commited and pushed that yesterday at the end of the day, keeping in head that I have to fix that today!
I think I'll strdup each result of ft_strnstr so I can free the entire char **imgs_name and don't have leaks issues anymore.
What do you think about that ?

**Comment by ZeanZarzin** [comment]
Yeah, that would work, but wouldn't it also be possible to edit the IMGS macro to just give the relative paths instead? That would mean it doesn't even have to ft_strnstr or anything.
Currently, if the whole project ended up inside a folder called something containing "sprites", it also wouldn't cut it correctly.

**Comment by aautin** [comment]
> Yeah, that would work, but wouldn't it also be possible to edit the IMGS macro to just give the relative paths instead? That would mean it doesn't even have to ft_strnstr or anything.> 

Yeah, I'm doing researchs about that. I used readlink here but I'm trying to find another command (like ls + options of something) to get the relative path.
Ok, I think I got something : `readlink -f sprites/* sprites/*/* | grep .xpm | awk -Fso_long/ '{print $2}'`
I added "`awk -Fso_long/ '{print $2}'`" to select the part after the so_long/. It seems to give me the result I needed.


> Currently, if the whole project ended up inside a folder called something containing "sprites", it also wouldn't cut it correctly.

In this project, I guess I don't have to handle this case 'cause it's mandated to call the project folder as "so_long".
I'm thinking about to add the subject of every project from 42School, so everyone might easily understand the limits and the rules of the programs

**Comment by aautin** [comment]
> Currently, if the whole project ended up inside a folder called something containing "sprites", it also wouldn't cut it correctly.

In this project, I guess I don't have to handle this case 'cause it's mandatory to call the project folder as "so_long".
I'm thinking about to add the subject of every project from 42School, so everyone might easily understand the limits and the rules of the programs
